stdlib/asm/crypto/hashes/rpo.masm

```diff
@@ -100,3 +100,155 @@ export.hash_memory
   # drop start_addr/end_addr (4 cycles)
   movup.4 drop movup.4 drop
 end
+
+#! Computes hash of Felt values starting at the specified memory address.
+#!
+#! If the number if inputs is 0, procedure returns the empty word: [ZERO, ZERO, ZERO, ZERO].
+#!
+#! Inputs:  [inputs_ptr, num_inputs]
+#! Outputs: [HASH]
+export.hash_memory
+    # move number of inputs to the top of the stack 
+    swap
+    # => [num_inputs, inputs_ptr]
+
+    # get the number of double words
+    u32divmod.8 swap
+    # => [num_inputs/8, num_inputs%8, inputs_ptr]
+
+    # get the end_addr for hash_memory_even procedure (end address for pairs of words)
+    mul.2 dup.2 add movup.2
+    # => [inputs_ptr, end_addr, num_inputs%8]
+
+    # get the capacity element which is equal to num_inputs%8
+    dup.2
+    # => [capacity, inputs_ptr, end_addr, num_inputs%8]
+    
+    # prepare hasher state for RPO permutation 
+    push.0.0.0 padw padw 
+    # => [C, B, A, inputs_ptr, end_addr, num_inputs%8]
+
+    # hash every pair of words
+    exec.absorb_double_words_from_memory
+    # => [C', B', A', inputs_ptr', end_addr, num_inputs%8] where inputs_ptr' = end_addr
+
+    # hash remaining input values if there are any left
+    # if num_inputs%8 is not ZERO
+    dup.14 push.0 neq
+    if.true
+        # load the remaining double word
+        mem_stream
+        # => [E, D, A', inputs_ptr'+2, end_addr, num_inputs%8]
+
+        # clean the stack
+        movup.12 drop movup.12 drop
+        # => [E, D, A', num_inputs%8]
+
+        # get the number of elements we need to drop
+        # notice that drop_counter could be any number from 1 to 7
+        push.8 movup.13 sub movdn.12
+        # => [E, D, A', drop_counter]
+
+        ### 0th value ########################################################
+        
+        # we need to drop first value anyway, since number of values is not divisible by 8
+        drop
+        # => [e_1, e_2, e_3, d_0, d_1, d_2, d_3, A', drop_counter]
+
+        # push the padding 0 value 
+        push.0
+        # => [0, e_1, e_2, e_3, d_0, d_1, d_2, d_3, A', drop_counter]
+
+        # move the calculated value down the stack 
+        movdn.6
+        # => [e_1, e_2, e_3, d_0, d_1, d_2, 0, d_3, A', drop_counter]
```

Comment:
The numbering of elements is opposite of what we usually do. Also, I'm not sure we benefit much from a more elaborate explanation here. I'd write this simply as:
```
drop push.0 movdn.6
# => [e_3, e_2, e_1, d_3, d_2, d_1, 0, d_0, A', drop_counter]
```